### DISCUSSION TITLE
Quality concerns with Flash Attention

### DISCUSSION THREAD

**Discussion by kaetemi** [ggml-org/llama.cpp#9646] [General]
Based on my own observations and user feedback for my application, I've noticed a significant degradation of output quality when flash attention is enabled. This was noticed on Llama3-8B and Falcon-7B base models, both running on GPU.

The application is continuously making use of context shifting. Is there any known interaction issue between this and flash attention? 

Particularly, when flash attention is enabled, completion becomes extremely vague or superficially reactionary and randomly switches topic.

Any similar experiences, or any thoughts on the combination of context shifting and flash attention?

**Comment by kaetemi** [comment]
Here's two attempts with flash attention enabled. The response questions are irrelevant, and the conversation completely loses focus.

![image](https://github.com/user-attachments/assets/2c71eeaa-4fde-4fc0-9ef8-a6b4ec2ee6ee)
![image](https://github.com/user-attachments/assets/e7a0c553-502d-4482-83fd-97f04c65e7ad)

And here's with flash attention *disabled*. Settings otherwise entirely the same. Guesses the right country in two questions.

![image](https://github.com/user-attachments/assets/413b11db-e879-4e72-b21b-7b7234626feb)

**Comment by kaetemi** [comment]
Tested on an NVIDIA RTX 4000 SFF Ada Generation, an RTX 3060 and a GTX 1080. All seem to be affected.

**Comment by OMS24** [comment]
It seems the way the implement it for older you than amper make it that its hamer the quality for speed

**Comment by l29ah** [comment]
Is it any good nowadays?

**Comment by MB7979** [comment]
I still see degradation. But it feels model specific. Sometimes the output is just different without being ‘bad’ different. Sometimes it’s plainly worse. Interesting choice in a recent commit to make flash attention be enabled by default.

**Reply by navr32** [reply]
Yes i have the same problem on RTX3090 . Better quality without flash attention on several models..

**Reply by Jacfger** [reply]
are you using lower bit like 16 or even 8/4 bits? In theory the lower bits the worse degradation

**Comment by subspecs** [comment]
Seriously, isn't it high time Flash Attention for GGML is finally fixed already?
Jesus christ I just tested it on the latest version and fuck its broken af.
It 'works'-ish, but after a few prompts it basically shits itself unless you do hello world tutorials.

I mean, we already have Flash Attention 2.0, Sage Attention 1, 2 and now even 3.
Can't we seriously come up with something good for GGML or at the very least fix the fucking Flash Attention?
It's a very useful tool since it increases speed and reduces VRAM, but what's the fucking point if it's buggy and broken.

I mean no disrespect here, but it's been many fucking years.
And there have been several people suggesting this, but all of it was rejected because "it would give no benefits etc." bs, when even now in its broken state it fucking GIVES vram/speed benefits already.

**Reply by 0cc4m** [reply]
It's working perfectly fine for me, and it's on by default for most users. If it were completely broken we'd be flooded with issues. If you have an actual measurable problem, please open an issue and post everything needed for reproduction.

**Reply by kaetemi** [reply]
Yep, this was fixed a long time ago already. It's the floating point precision issue

**Reply by subspecs** [reply]
Alright. I'll make another discussion with measurements to illustrate my point. It IS better than years ago, but still has issues.

**Reply by 0cc4m** [reply]
discussions are not really tracked by maintainers, issues is the right place to report (surprise) issues.
